rust/lance-io/src/scheduler.rs

```diff
@@ -258,8 +270,61 @@ impl FileScheduler {
     ) -> impl Future<Output = Result<Vec<Bytes>>> + Send {
         // The final priority is a combination of the row offset and the file number
         let priority = ((self.file_index as u128) << 64) + priority as u128;
-        self.root
-            .submit_request(self.reader.clone(), request, priority)
+
+        let mut updated_requests = Vec::with_capacity(request.len());
+
+        let copy_request = request.clone();
+
+        if !request.is_empty() {
+            let mut curr_interval = request[0].clone();
+
+            for req in request.iter().skip(1) {
+                if is_close_together(&curr_interval, req, self.block_size) {
+                    curr_interval.end = curr_interval.end.max(req.end);
+                } else {
```

Comment:
Doesn't this assume the requests are already kind of monotonic? Is that handled anywhere? Should we sort the requests by `request.start` before doing this?

Reply:
Yes, I believe requests (`ranges`) are assumed to be sorted by a few encodings at this point. I don't think this is enforced anywhere though. We should probably make a note of wherever this sorting is assumed.
It may be better to sort the requests in the encoding itself (inside `schedule_ranges()`), before they are passed to `submit_request()` since then we won't have to call sort multiple times. In that case it may be better to leave this for another PR?

Reply:
I see. Looks like you are coalescing on the request side, and so they are all from the same array. If we put this one a different side, where they might be sourced from multiple decoders, it might be different. Seems fine for now, but wanted to note that seems to be an implicit assumption.

Reply:
Yes, I broke coalescing up into two github issues: "in batch coalescing" which is cheap (we can assume all requests are for a single page of data and they arrive in sorted order) and "out of batch coalescing".  I'm not sure we ever want to bother with "out of batch coalescing".  You'd need some kind of sorted heap of requests and then you'd have to unsort them after the request is made.  You'd also get weird things like what happens if you can coalesce part of a low priority request that's way back in the queue with a high priority request that's almost ready to run.  You'd also need to make sure you aren't blocking the I/O threads at any point while you sort your queue.  :skull: